### DISCUSSION TITLE
FR: One-click clear context

### DISCUSSION THREAD

**Discussion by fangyuan99** [open-webui/open-webui#5971] [General]
When using other projects of the same type, they all provide a one-click context clearing button, so there is no need to start a new conversation

**Comment by rgaricano** [comment]
This kind of thing can be done with tools and functions, some are already ready: https://openwebui.com/functions/?query=context , they can be installed and managed through the chat workplace and used from the chat input.
(summarize, limit, cut...context).

**Comment by rgaricano** [comment]
I've modified that filter to work as an action and be able to activate it at will. Now add a button in the last message to cut the context as defined in the function valve (in the function configuration) (by default 1), create a function, paste code, activate & activate it as global as well (in three points of the function configuration)
In the end I haven't added a confirmation message, when I have a little time I'll improve it,..., or maybe you can do it.

(edited: added lost lib)
(edited1: DOESN'T WORK, JSON PACKAGE SENT CONTAINS ALL MESSAGES )

```
"""
title: Context Cutter Action
author: open-webui
author_url: https://github.com/open-webui
funding_url: https://github.com/open-webui
version: 0.1
"""

from pydantic import BaseModel, Field
from typing import Optional
from open_webui.apps.webui.models.users import Users

class Action:
    class Valves(BaseModel):
        priority: int = Field(
            default=0, description="Priority level for the filter operations."
        )
        n_last_messages: int = Field(
            default=1, description="Number of last messages to retain."
        )
        pass

    class UserValves(BaseModel):
        show_status: bool = Field(
            default=True, description="Show status of the action."
        )

    pass

    def __init__(self):
        self.valves = self.Valves()
        pass

    async def action(
        self,
        body: dict,
        __user__=None,
        __event_emitter__=None,
        __event_call__=None,
    ) -> Optional[dict]:
        if not __event_emitter__:
            return None
        print(f"action:{__name__}")

        user_valves = __user__.get("valves")
        if not user_valves:
            user_valves = self.UserValves()
        await __event_emitter__(
            {
                "type": "status",
                "data": {
                    "description": "Analysing Data",
                    "done": False,
                },
            }
        )

        if __event_emitter__:
            messages = body["messages"]  # Ensure we always keep the system prompt
            user = Users.get_user_by_id(__user__["id"])

            if user_valves.show_status:
                await __event_emitter__(
                    {
                        "type": "status",
                        "data": {"description": "Cutting context", "done": False},
                    }
                )

            # add the assistant response to memories
            try:
                system_prompt = next(
                    (
                        message
                        for message in messages
                        if message.get("role") == "system"
                    ),
                    None,
                )
                if system_prompt:
                    messages = [
                        message
                        for message in messages
                        if message.get("role") != "system"
                    ]
                    messages = messages[-self.valves.n_last_messages :]
                    messages.insert(0, system_prompt)
                else:  # If no system prompt, simply truncate to the last n_last_messages
                    messages = messages[-self.valves.n_last_messages :]
                body["messages"] = messages
                print(f"Context cutted to last: {n_last_messages} messages")
            except Exception as e:
                print(f"Error cutting context {str(e)}")
                if user_valves.show_status:
                    await __event_emitter__(
                        {
                            "type": "status",
                            "data": {
                                "description": "Error cutting context",
                                "done": True,
                            },
                        }
                    )

                    # add a citation to the message with the error
                    await __event_emitter__(
                        {
                            "type": "citation",
                            "data": {
                                "source": {"name": "Error:cutting context"},
                                "document": [str(e)],
                                "metadata": [
                                    {"source": "Add to cutting context Action Button"}
                                ],
                            },
                        }
                    )

            if user_valves.show_status:
                await __event_emitter__(
                    {
                        "type": "status",
                        "data": {"description": "cutting context done", "done": True},
                    }
                )

```

**Reply by fangyuan99** [reply]
Thanks for your reply, but I tried it and didn't find the cut button.
I don't know if you've used https://v3.nextchat.dev/#/ ,it has a button to clear and restore previous memories with one click, very handy!
![image](https://github.com/user-attachments/assets/c3910a32-7183-44bc-93fc-085d6ed66d83)

**Reply by rgaricano** [reply]
under previous message box, it¡s necessary to activate as global on function setup (workspace-functions-... of defined function)

![cut_context_button](https://github.com/user-attachments/assets/cd6d6acd-27fa-4909-bb1a-508891111602)
![cut_context_option](https://github.com/user-attachments/assets/11828b81-6f61-4013-9540-ef142d861d95)

**Reply by fangyuan99** [reply]
The button appeared but an error occurred. I'll try to debug it
![image](https://github.com/user-attachments/assets/824fe602-fd7d-4452-9651-785232329b15)

**Reply by rgaricano** [reply]
I haven´t that error, try adding 
`from open_webui.apps.webui.models.users import Users`

**Reply by fangyuan99** [reply]
![image](https://github.com/user-attachments/assets/7c280210-4d17-4cfe-8716-db3e40a69028)
It doesn't seem to work, I captured the request package and found that it still carries all the context.

**Reply by rgaricano** [reply]
I'll try to look at it later, I can't now, keep in mind that this is a quick adaptation as an example and my knowledge of openwebui is very limited, anyway by default this action retains 1 message...

**Comment by rgaricano** [comment]
Another easy solution is tell to  IA that delete all previous context, it do!

**Reply by fangyuan99** [reply]
But it still consumes tokens, right?

**Comment by justinh-rahb** [comment]
Isn't this effectively just starting a new chat with extra steps? Genuinely would like to understand the usefulness of this, right now I just don't see it.

**Reply by rgaricano** [reply]
Me neither, but at least it helps me practice with functions.

**Reply by fangyuan99** [reply]
If you create a new chat, I can't keep the current model and I can easily check the information of the previous conversation

**Comment by oatmealm** [comment]
If there's a collection and model configuration that the user wants to keep, this could be very useful, if the reset action only cleans the context up.

**Reply by rgaricano** [reply]
I've been looking around a bit and I can't find the key.
The filter works, but if it's with an action, it doesn't. This is logical, since the filter is applied when sending, while the action doesn't. This should modify the message array or the chat history variable, and I, for now, don't know how to do it without creating inconsistencies.
Another possibility that I've seen is that the action would create a new chat with a different id, so the history and the message array would be initialized, but I don't see it clearly.
The problem with the filter is that it always cuts the context in n messages.
I don't know the code well enough, but from what I see, if we modify the history or the message count, it would also affect the display itself.
(for ref. scr/lib/components/chat/Messages.svelte & Chat.svelte )
